lib/ResourceSet.h

```diff
@@ -32,66 +33,146 @@ namespace Res
 	};
 
 	//class to be representing a vector of resource
-	class ResourceSet : public std::vector<int>
+	class ResourceSet
 	{
+	private:
+		std::array<int, GameConstants::RESOURCE_QUANTITY> container;
 	public:
-		DLL_LINKAGE ResourceSet();
 		// read resources set from json. Format example: { "gold": 500, "wood":5 }
 		DLL_LINKAGE ResourceSet(const JsonNode & node);
-		DLL_LINKAGE ResourceSet(TResource wood, TResource mercury, TResource ore, TResource sulfur, TResource crystal,
-								TResource gems, TResource gold, TResource mithril = 0);
+		DLL_LINKAGE ResourceSet(TResource wood = 0, TResource mercury = 0, TResource ore = 0, TResource sulfur = 0, TResource crystal = 0,
+								TResource gems = 0, TResource gold = 0, TResource mithril = 0);
 
 
 #define scalarOperator(OPSIGN)									\
-		ResourceSet operator OPSIGN(const TResource &rhs) const	\
+		ResourceSet& operator OPSIGN ## =(const TResource &rhs) \
 		{														\
-			ResourceSet ret = *this;							\
-			for(int i = 0; i < (int)size(); i++)						\
-				ret[i] = at(i) OPSIGN rhs;						\
+			for(auto i = 0; i < container.size(); i++)						\
+				at(i) OPSIGN ## = rhs;						\
 																\
-			return ret;											\
+			return *this;											\
 		}
 
-
-
 #define vectorOperator(OPSIGN)										\
-		ResourceSet operator OPSIGN(const ResourceSet &rhs) const	\
+		ResourceSet& operator OPSIGN ## =(const ResourceSet &rhs)	\
 		{															\
-			ResourceSet ret = *this;								\
-			for(int i = 0; i < (int)size(); i++)							\
-				ret[i] = at(i) OPSIGN rhs[i];						\
+			for(auto i = 0; i < container.size(); i++)							\
+				at(i) OPSIGN ## = rhs[i];						\
 																	\
-			return ret;												\
+			return *this;												\
 		}
 
-
-#define opEqOperator(OPSIGN, RHS_TYPE)							\
-		ResourceSet& operator OPSIGN ## =(const RHS_TYPE &rhs)	\
-		{														\
-			return *this = *this OPSIGN rhs;					\
+#define twoOperands(OPSIGN, RHS_TYPE) \
+		friend ResourceSet operator OPSIGN(ResourceSet lhs, const RHS_TYPE &rhs) \
+		{ \
+			lhs OPSIGN ## = rhs; \
+			return lhs; \
 		}
 
 		scalarOperator(+)
 		scalarOperator(-)
 		scalarOperator(*)
 		scalarOperator(/)
-		opEqOperator(+, TResource)
-		opEqOperator(-, TResource)
-		opEqOperator(*, TResource)
 		vectorOperator(+)
 		vectorOperator(-)
-		opEqOperator(+, ResourceSet)
-		opEqOperator(-, ResourceSet)
+		twoOperands(+, TResource)
+		twoOperands(-, TResource)
+		twoOperands(*, TResource)
+		twoOperands(/, TResource)
+		twoOperands(+, ResourceSet)
+		twoOperands(-, ResourceSet)
+
 
 #undef scalarOperator
 #undef vectorOperator
-#undef opEqOperator
+#undef twoOperands
+
+		using const_reference = decltype(container)::const_reference;
+		using value_type = decltype(container)::value_type;
+		using const_iterator = decltype(container)::const_iterator;
+		using iterator = decltype(container)::iterator;
+
+		// Array-like interface
+		TResource & at(Res::ERes index)

```

Comment:
Do we need at() here? Perhaps always use operator[]?

Reply:
@IvanSavenko what is wrong with at?

Reply:
Don't see any need for it. Currently this only leads to more largely identical methods in resource set.

We almost never use at() in code in general, with exception of const std::map that does not have operator[].

Reply:
at removed.

Reply:
Removal of .at breaks compilation in many places. Maybe save at?